XenModel/Actions/Updates/DownloadAndUnzipXenServerPatchAction.cs

```diff
@@ -329,7 +338,35 @@ void client_DownloadFileCompleted(object sender, AsyncCompletedEventArgs e)
 
             if (e.Error != null) //failure
             {
-                patchDownloadError = e.Error;
+                if (e.Error is WebException wex && wex.Response is HttpWebResponse response)
+                {
+                    switch (response.StatusCode)
+                    {
+                        case HttpStatusCode.Unauthorized:
+                            log.Error($"Could not download {updateName} (401 Unauthorized). Client ID may be invalid or revoked.");
+                            patchDownloadError = new Exception(Messages.FILESERVICE_ERROR_401);
+                            break;
+
+                        case HttpStatusCode.Forbidden:
+                            log.Error($"Could not download {updateName} (403 Forbidden). The account has insufficient permissions.");
+                            patchDownloadError = new Exception(Messages.FILESERVICE_ERROR_403);
+                            break;
+
+                        case HttpStatusCode.NotFound:
+                            log.Error($"Could not download {updateName} (404 File not found).");
+                            patchDownloadError = new Exception(Messages.FILESERVICE_ERROR_404);
+                            break;
+
+                        default:
+                            patchDownloadError = e.Error;

```

Comment:
Shouldn't we log even if its not the above statuses

Reply:
These are caught and logged further up (see method DownloadFile).
Throughtout XenModel actions you may see that some things are caught and logged within the Run() method (which is the action body) when we want to show a specialized message. What is not handled there is always caught within the base AsyncAction or ActionBase classes.